lib/WP_Auth0_LoginManager.php

```diff
@@ -163,136 +170,124 @@ public function init_auth0() {
+        __( 'An oauth/token call triggered a 401 response from Auth0. ', 'wp-auth0' ) .
+          __( 'Please check the Client Secret saved in the Auth0 plugin settings. ', 'wp-auth0' )
+      );
+      throw new WP_Auth0_LoginFlowValidationException( __( 'Not Authorized', 'wp-auth0' ), $exchange_resp_code );
 
-	    $decoded_token = JWT::decode(
-		    $data->id_token,
-		    $this->a0_options->get_client_secret_as_key(),
-		    array( $this->a0_options->get_client_signing_algorithm() )
-	    );
+    } else if ( empty( $exchange_resp_body ) ) {
 
-	    $data->id_token = null;
-	    $response = WP_Auth0_Api_Client::get_user(
-		    $this->a0_options->get( 'domain' ),
-		    WP_Auth0_Api_Client::get_client_token(),
-		    $decoded_token->sub
-	    );
+      // Unsuccessful for another reason
+      if ( $exchange_resp instanceof WP_Error ) {
+        WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' L:' . __LINE__, $exchange_resp );
+      }
 
-      if ( $response instanceof WP_Error ) {
-        WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => WP_Auth0_Api_Client::get_user()', $response );
+      throw new WP_Auth0_LoginFlowValidationException( __( 'Unknown error', 'wp-auth0' ), $exchange_resp_code );
+    }
 
-        error_log( $response->get_error_message() );
+    $data = json_decode( $exchange_resp_body );
 
-        throw new WP_Auth0_LoginFlowValidationException( );
-      }
+    if ( empty( $data->access_token ) ) {
 
-      $userinfo = json_decode( $response['body'] );
-      if ( $this->login_user( $userinfo, $data->id_token, $data->access_token ) ) {
-        if ( ! empty( $state_decoded->interim ) ) {
-          include WPA0_PLUGIN_DIR . 'templates/login-interim.php';
-          exit();
-        } else {
-          if ( ! empty( $state_decoded->redirect_to ) && wp_login_url() !== $state_decoded->redirect_to ) {
-            $redirectURL = $state_decoded->redirect_to;
-          } else {
-            $redirectURL = $this->a0_options->get( 'default_login_redirection' );
-          }
+      // Look for clues as to what went wrong
+      throw new WP_Auth0_LoginFlowValidationException(
+        ! empty( $data->error_description ) ? $data->error_description : __( 'Unknown error', 'wp-auth0' ),
+        ( ! empty( $data->error ) ? $data->error : $exchange_resp_code )
+      );
+    }
 
-          wp_safe_redirect( $redirectURL );
-        }
-      }
-    } elseif ( is_array( $response['response'] ) &&  401 === (int) $response['response']['code'] ) {
+    $decoded_token = JWT::decode(
+      $data->id_token,
+      $this->a0_options->get_client_secret_as_key(),
+      array( $this->a0_options->get_client_signing_algorithm() )
+    );
 
-      $error = new WP_Error( '401', 'auth/token response code: 401 Unauthorized' );
+    $userinfo_resp = WP_Auth0_Api_Client::get_user(
+      $this->a0_options->get( 'domain' ),
+      WP_Auth0_Api_Client::get_client_token(),
+      $decoded_token->sub
```

Comment:
Same as a comment I made previously. What happens if `$decoded_token` is null/undefined because the decode function failed. Will this create a "request" anyway and execute it? if true, is not resource friendly since we know in advance it failed so the request will fail as well.

Reply:
Good call, a bit of a shortcut on my part. Fix OTW.